Fix Http3ConnectionEmptyBenchmark

amcasey opened the pull request:
It was reusing an enumerator that had already been enumerated.  The enumerator doesn't support Reset, so this change just reinitializes it.

amcasey committed:
Fix Http3ConnectionEmptyBenchmark

It was reusing an enumerator that had already been enumerated.  The enumerator doesn't support Reset, so this change just reinitializes it.

amcasey commented:
FYI @BrennanConroy, with whom I've been discussing this offline.

halter73 commented on src/Servers/Kestrel/perf/Microbenchmarks/Http3/Http3ConnectionBenchmarkBase.cs:
```diff
@@ -72,6 +72,8 @@ public async Task MakeRequest()
 
+        _requestHeadersEnumerator.Initialize(_httpRequestHeaders);
+
         await stream.SendHeadersAsync(_requestHeadersEnumerator);
```
Doesn't `_http3.CreateRequestStream` already call `SendHeadersAsync`? I wonder why this test was calling it twice. We probably shouldn't re-enumerate it though.

https://github.com/dotnet/aspnetcore/blob/600eb9aa53c052ec7327e2399744215dbe493a89/src/Servers/Kestrel/shared/test/Http3/Http3InMemory.cs#L445

Maybe we should just delete this line? Do we get a response if we don't call `SendHeadersAsync` again?

amcasey replied:
With it removed, we now get a single 45-byte frame, rather than an immediate null (i.e. I'm pretty sure it was broken).

amcasey committed:
Drop superfluous SendHeadersAsync call

@halter73 correctly pointed out that `CreateRequestStream` already calls `SendHeadersAsync`.

halter73 reviewed (approved):
Out of curiosity, did this change the results? I imagine it wouldn't have a big impact since the mutable enumerator was already enumerated, so the headers probably weren't actually resent.

amcasey commented:
> Out of curiosity, did this change the results? I imagine it wouldn't have a big impact since the mutable enumerator was already enumerated, so the headers probably weren't actually resent.

@halter73 How would I check?

amcasey commented:
Oh, sorry, I think that was the fault in my previous repo, so I didn't even look.  Can/should it be the default here?

halter73 commented:
GitHub saves whatever you did last in the repo.
